### ISSUE TITLE
Integration has failed to load

### ISSUE THREAD

**Issue by hilburn** [h4de5/home-assistant-toshiba_ac#207]
About 10 hours ago, Toshiba AC stopped providing all the data on my devices
Now trying to access the climates for each AC unit I get "unavailable" and in settings:
![image](https://github.com/user-attachments/assets/5d355681-cf8b-41ab-a9f4-aff85ef1a2f2)

It corresponds with the following in the logs:
```
2024-10-21 23:30:06.014 WARNING (SyncWorker_0) [homeassistant.loader] We found a custom integration toshiba_ac which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2024-10-21 23:30:23.175 WARNING (MainThread) [homeassistant.components.climate] Setup of climate platform toshiba_ac is taking over 10 seconds.
2024-10-21 23:30:23.177 WARNING (MainThread) [homeassistant.components.select] Setup of select platform toshiba_ac is taking over 10 seconds.
2024-10-21 23:30:23.177 WARNING (MainThread) [homeassistant.components.sensor] Setup of sensor platform toshiba_ac is taking over 10 seconds.
2024-10-21 23:30:23.177 WARNING (MainThread) [homeassistant.components.switch] Setup of switch platform toshiba_ac is taking over 10 seconds.
2024-10-21 23:31:06.628 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: {('toshiba_ac', '52079250f415eef76f33a386c395c852'): 47.530426356}
2024-10-21 23:31:13.177 ERROR (MainThread) [homeassistant.components.climate] Setup of platform toshiba_ac is taking longer than 60 seconds. Startup will proceed without waiting any longer.
2024-10-21 23:31:13.177 ERROR (MainThread) [homeassistant.components.select] Setup of platform toshiba_ac is taking longer than 60 seconds. Startup will proceed without waiting any longer.
2024-10-21 23:31:13.177 ERROR (MainThread) [homeassistant.components.sensor] Setup of platform toshiba_ac is taking longer than 60 seconds. Startup will proceed without waiting any longer.
2024-10-21 23:31:13.177 ERROR (MainThread) [homeassistant.components.switch] Setup of platform toshiba_ac is taking longer than 60 seconds. Startup will proceed without waiting any longer.
2024-10-21 23:33:08.121 ERROR (MainThread) [custom_components.toshiba_ac.climate] Error during connection to Toshiba server
File "/config/deps/lib/python3.12/site-packages/toshiba_ac/utils/__init__.py", line 90, in wrapper
File "/config/deps/lib/python3.12/site-packages/toshiba_ac/utils/http_api.py", line 114, in request_api
raise ToshibaAcHttpApiError(await response.text())
toshiba_ac.utils.http_api.ToshibaAcHttpApiError: <html>
File "/config/deps/lib/python3.12/site-packages/toshiba_ac/utils/__init__.py", line 96, in wrapper
File "/config/custom_components/toshiba_ac/climate.py", line 52, in async_setup_entry
File "/config/deps/lib/python3.12/site-packages/toshiba_ac/device_manager.py", line 191, in get_devices
File "/config/deps/lib/python3.12/site-packages/toshiba_ac/device/__init__.py", line 88, in connect
File "/config/deps/lib/python3.12/site-packages/toshiba_ac/device/__init__.py", line 97, in load_additional_device_info
File "/config/deps/lib/python3.12/site-packages/toshiba_ac/utils/http_api.py", line 177, in get_device_additional_info
File "/config/deps/lib/python3.12/site-packages/toshiba_ac/utils/__init__.py", line 63, in wrapper
File "/config/custom_components/toshiba_ac/climate.py", line 58, in async_setup_entry
raise ConfigEntryNotReady("Error during connection to Toshiba server") from ex
homeassistant.exceptions.ConfigEntryNotReady: Error during connection to Toshiba server
2024-10-22 02:33:07.006 ERROR (MainThread) [toshiba_ac.device] State reload failed:
2024-10-22 03:32:41.553 ERROR (MainThread) [toshiba_ac.device] State reload failed:
2024-10-22 06:04:42.578 ERROR (MainThread) [toshiba_ac.device] State reload failed:
2024-10-22 07:03:05.416 ERROR (MainThread) [toshiba_ac.device] State reload failed:
2024-10-22 09:07:20.339 ERROR (MainThread) [custom_components.toshiba_ac] Unload Toshiba integration
2024-10-22 09:07:32.479 ERROR (MainThread) [custom_components.toshiba_ac.config_flow] Toshiba connection OK
```

**Comment by jalada** [comment]
If it helps debug, I'm also having problems but I have slightly different logs that contain the full error from their API, which is a 403 Forbidden from their Azure API gateway:

```
2024-10-29 06:03:56.841 ERROR (MainThread) [custom_components.toshiba_ac.climate] Error during connection to Toshiba server
2024-10-29 06:03:56.842 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved (None)
Traceback (most recent call last):
File "/usr/local/lib/python3.12/site-packages/toshiba_ac/utils/__init__.py", line 90, in wrapper
return await func(*args, **kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/toshiba_ac/utils/http_api.py", line 114, in request_api
raise ToshibaAcHttpApiError(await response.text())
toshiba_ac.utils.http_api.ToshibaAcHttpApiError: <html>
<head><title>403 Forbidden</title></head>
<body>
<center><h1>403 Forbidden</h1></center>
<hr><center>Microsoft-Azure-Application-Gateway/v2</center>
</body>
</html>
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
File "/usr/local/lib/python3.12/asyncio/tasks.py", line 520, in wait_for
return await fut
^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/toshiba_ac/utils/__init__.py", line 96, in wrapper
await asyncio.sleep(bk)
File "/usr/local/lib/python3.12/asyncio/tasks.py", line 665, in sleep
return await future
^^^^^^^^^^^^
asyncio.exceptions.CancelledError
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
File "/config/custom_components/toshiba_ac/climate.py", line 52, in async_setup_entry
devices = await device_manager.get_devices()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/toshiba_ac/device_manager.py", line 191, in get_devices
await asyncio.gather(*connects)
File "/usr/local/lib/python3.12/site-packages/toshiba_ac/device/__init__.py", line 88, in connect
await self.load_additional_device_info()
File "/usr/local/lib/python3.12/site-packages/toshiba_ac/device/__init__.py", line 97, in load_additional_device_info
additional_info = await self.http_api.get_device_additional_info(self.ac_id)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/toshiba_ac/utils/http_api.py", line 177, in get_device_additional_info
res = await self.request_api(self.AC_STATE_PATH, get=get)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/toshiba_ac/utils/__init__.py", line 63, in wrapper
return await asyncio.wait_for(func(*args, **kwargs), timeout=timeout)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/asyncio/tasks.py", line 519, in wait_for
async with timeouts.timeout(timeout):
File "/usr/local/lib/python3.12/asyncio/timeouts.py", line 115, in __aexit__
raise TimeoutError from exc_val
TimeoutError
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
File "/config/custom_components/toshiba_ac/climate.py", line 58, in async_setup_entry
raise ConfigEntryNotReady("Error during connection to Toshiba server") from ex
homeassistant.exceptions.ConfigEntryNotReady: Error during connection to Toshiba server
```

**Comment by jalada** [comment]
@hilburn I actually fixed this by removing and re-adding the integration - the 403 Forbidden I was getting was a clue that maybe a re-auth was needed.

Your logs look like they have the full response chopped off:

> `toshiba_ac.utils.http_api.ToshibaAcHttpApiError: <html>`

Do you know if there were other lines, containing the full contents of the response?

**Comment by albaintor** [comment]
Hi, there is indeed a problem with startups. I know that you added code when the authentication token expires or no longer works. However this logic does not work when HA restarts.
At (almost) each restart of HA (after a core or component update for ex) : the integration will fail to connect to Toshiba servers and will not try to renew the token. I have to relaunch HA again to make this work.
The disable/enable integration won't trigger any renewal of the token, so only restart of HA (or delete/add integration again) will fix the problem.

Something has to be fixed in the initialization of the component, also you could add an option flow to reset the credentials. I can help on that

**Comment by albaintor** [comment]
A few precisions :
`[homeassistant.bootstrap] Waiting on integrations to complete setup: {('toshiba_ac', 'XXXXXXXXX'): 285742.60670832}`

Then this error message :
`
Error doing job: Task exception was never retrieved (None)
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/toshiba_ac/utils/__init__.py", line 90, in wrapper
    return await func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/toshiba_ac/utils/http_api.py", line 114, in request_api
    raise ToshibaAcHttpApiError(await response.text())
toshiba_ac.utils.http_api.ToshibaAcHttpApiError: <html>
<head><title>403 Forbidden</title></head>
<body>
<center><h1>403 Forbidden</h1></center>
<hr><center>Microsoft-Azure-Application-Gateway/v2</center>
</body>
</html>


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/asyncio/tasks.py", line 520, in wait_for
    return await fut
           ^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/toshiba_ac/utils/__init__.py", line 96, in wrapper
    await asyncio.sleep(bk)
  File "/usr/local/lib/python3.12/asyncio/tasks.py", line 665, in sleep
    return await future
           ^^^^^^^^^^^^
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/config/custom_components/toshiba_ac/climate.py", line 52, in async_setup_entry
    devices = await device_manager.get_devices()
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/toshiba_ac/device_manager.py", line 191, in get_devices
    await asyncio.gather(*connects)
  File "/usr/local/lib/python3.12/site-packages/toshiba_ac/device/__init__.py", line 88, in connect
    await self.load_additional_device_info()
  File "/usr/local/lib/python3.12/site-packages/toshiba_ac/device/__init__.py", line 97, in load_additional_device_info
    additional_info = await self.http_api.get_device_additional_info(self.ac_id)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/toshiba_ac/utils/http_api.py", line 177, in get_device_additional_info
    res = await self.request_api(self.AC_STATE_PATH, get=get)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/toshiba_ac/utils/__init__.py", line 63, in wrapper
    return await asyncio.wait_for(func(*args, **kwargs), timeout=timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/tasks.py", line 519, in wait_for
    async with timeouts.timeout(timeout):
  File "/usr/local/lib/python3.12/asyncio/timeouts.py", line 115, in __aexit__
    raise TimeoutError from exc_val
TimeoutError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/config/custom_components/toshiba_ac/climate.py", line 58, in async_setup_entry
    raise ConfigEntryNotReady("Error during connection to Toshiba server") from ex
homeassistant.exceptions.ConfigEntryNotReady: Error during connection to Toshiba server`

**Comment by felixmueller** [comment]
@albaintor I am experiencing the exact same issue: only after a second restart of ha core, the token is renewed and the toshiba ac integration is working again. manual reload does not work.

**Comment by albaintor** [comment]
Ok this is more complicated than I thought : the component uses a third party library which performs 3 retries with a variable delay between retries (which can takes up to 60 seconds between each). See the decorator retry_with_timeout.

This is not the way to go when an HTTP 403 (unauthorized) error is raised. This error type should be filtered and directly returned to the caller. Then the caller (the HA component) should intercept 403 and regenerate the token

```python
@retry_with_timeout(timeout=5, retries=3, backoff=60)
    @retry_on_exception(exceptions=ToshibaAcHttpApiError, retries=3, backoff=60)
    async def request_api(
        self,
        path: str,
        get: dict[str, str] | None = None,
        post: t.Mapping[str, str | t.Sequence[str]] | None = None,
        headers: t.Any = None,
    ) -> t.Any:
```

**Comment by albaintor** [comment]
I have made further investigations. I don't understand what is going on with the server. I understand that there 2 endpoints, one to toshiba's server, and another one to microsoft azure. The problem is not related to Azure, which is where your logic against the token renewal applies. So as the SAS token seems to be okay, no renewal is necessary.

No, the problem is before :
Workflow :
1. First the connection to API to connect with user/password (POST /login)=> an access token is returned (and this step is working). This is not the same token as azure's (SAS) token.
2. Retrieval of the list of devices with the access token and consumerId returned in previous step (GET /GetConsumerACMapping) => this step is still okay, a list of devices is returned
3. Loop with extraction of each device information : this step fails (403 error) although this is the exact same access token

**Comment by albaintor** [comment]
Update : I have submitted the PR on the project of the creator of  the library. He has to approve it and release it. Then this integration will be able to be updated and fixed

**Comment by albaintor** [comment]
FYI : the developer didn't answer yet to my PR. Meantime, you can take the archive from this link https://github.com/albaintor/home-assistant-toshiba_ac/archive/refs/heads/fix_rejects.zip and replace files from subfolder `home-assistant-toshiba_ac-fix_rejects\custom_components\toshiba_ac\*` to home assistant following folder : `config/custom_components/toshiba_ac`
Restart HA and voila

**Comment by felixmueller** [comment]
@albaintor Hi! I just extracted your fix_rejects.zip into my current working directory (main branch: version 2024.9.0) and I cannot see any changed files. They all seem to match the current version. Are you sure you changed anything? Thanks!

**Comment by albaintor** [comment]
> @albaintor Hi! I just extracted your fix_rejects.zip into my current working directory (main branch: version 2024.9.0) and I cannot see any changed files. They all seem to match the current version. Are you sure you changed anything? Thanks!

Indeed, I don't know what happend, my commit didn't get through.
Here is the release with the right files : https://github.com/albaintor/home-assistant-toshiba_ac/releases/download/2024.9.0/toshiba_ac-2024.9.0.zip

**Comment by albaintor** [comment]
The external library has just been merged and updated with the PR : I have just submitted this PR now for the HA component
https://github.com/h4de5/home-assistant-toshiba_ac/pull/214

**Comment by h4de5** [comment]
fixed in #214
